Http/complete settings (#2972)
quicwg/base-drafts: draft-ietf-quic-http.md

#### Initialization

An HTTP implementation MUST NOT send frames or requests which would be invalid
based on its current understanding of the peer's settings.

All settings begin at an initial value.  Each endpoint MAY use these initial
values to send messages before the peer's SETTINGS frame has arrived.  When
the SETTINGS frame arrives, any settings are changed to their new values.

This removes the need to wait for the SETTINGS frame before sending messages.
Endpoints MUST NOT require any data to be received from the peer prior to
sending the SETTINGS frame; settings SHOULD be sent as soon as the transport is

> ready to send data.

Reviewer: I think it's worth adding an Implementation Note along the lines of:
"For clients using a 1-RTT QUIC connection, the 1-RTT keys will always become available prior to SETTINGS arriving, even if the server send SETTINGS immediately.  Clients SHOULD NOT wait indefinitely for SETTINGS to arrive before sending requests, but SHOULD process received datagrams in order to increase the likelihood of processing SETTINGS before sending the first request."

values to send messages before the peer's SETTINGS frame has arrived.  When
the SETTINGS frame arrives, any settings are changed to their new values.

This removes the need to wait for the SETTINGS frame before sending messages.
Endpoints MUST NOT require any data to be received from the peer prior to
sending the SETTINGS frame; settings SHOULD be sent as soon as the transport is
ready to send data.

For servers, the initial value of each client setting is the default value.

For clients using a 1-RTT QUIC connection, the initial value of each server
setting is the default value.  1-RTT keys will always become available prior to
SETTINGS arriving, even if the server sends SETTINGS immediately. Clients SHOULD
NOT wait indefinitely for SETTINGS to arrive before sending requests, but SHOULD
process received datagrams in order to increase the likelihood of processing
SETTINGS before sending the first request.

When a 0-RTT QUIC connection is being used, the initial value of each server
setting is the value used in the previous session. Clients SHOULD store the
settings the server provided in the connection where resumption information was

> provided, but MAY opt not to store settings in certain cases (e.g. if the

Reviewer 2: ```suggestion
provided, but MAY opt not to store settings in certain cases (e.g., if the
```

session ticket is received before the SETTINGS frame). A client MUST comply with
stored settings -- or default values, if no values are stored -- when attempting
0-RTT. Once a server has provided new settings, clients MUST comply with those
values.

A server can remember the settings that it advertised, or store an
integrity-protected copy of the values in the ticket and recover the information
when accepting 0-RTT data. A server uses the HTTP/3 settings values in
determining whether to accept 0-RTT data.  If the server cannot determine that
the settings remembered by a client are compatible with its current settings, it

> MUST NOT accept 0-RTT data.  (Remembered settings are compatible if a client

Reviewer 2: ```suggestion
MUST NOT accept 0-RTT data.  Remembered settings are compatible if a client
```

> complying with those settings would not violate the server's current settings.)

Reviewer 2: ```suggestion
complying with those settings would not violate the server's current settings.
```

A server MAY accept 0-RTT and subsequently provide different settings in its
SETTINGS frame. If 0-RTT data is accepted by the server, its SETTINGS frame MUST
NOT reduce any limits or alter any values that might be violated by the client
with its 0-RTT data.  The server MUST include all settings which differ from
their default values.  If a server accepts 0-RTT, but then sends a SETTINGS
frame which reduces a setting the client understands or omits a value that was

> previous specified, this MUST be treated as a connection error of type

Reviewer 2: ```suggestion
previously specified to have a non-default value, this MUST be treated as a connection error of type
```

The "ly" is just grammar, but I think that it would be OK to have a server send a default value in connection 1 and omit the value in connection 2.

> connection error of type HTTP_SETTINGS_ERROR.

Reviewer 3: a bit tangential but the description of HTTP_SETTINGS_ERROR doesn't mention this case

Reviewer 3: Sgtm

An HTTP implementation MUST NOT send frames or requests which would be invalid
based on its current understanding of the peer's settings.  All settings begin
at an initial value.  Each endpoint can use these values to send messages before
the peer's SETTINGS frame has arrived.  This removes the need to wait for the
SETTINGS frame before sending messages.  When the SETTINGS frame arrives, any
settings are changed to their new values.

For servers, the initial value of each client setting is the default value.

For clients using a 1-RTT QUIC connection, the initial value of each server
setting is the default value.  When a 0-RTT QUIC connection is being used, the
initial value of each server setting is the value used in the previous session.
Clients SHOULD store the settings the server provided in the connection where
resumption information was provided. A client MUST comply with either stored
settings -- or default values, if no values are stored -- when attempting 0-RTT.

> Once a server has provided new settings, clients MUST comply with those values.

Reviewer 4: This does smooth over the worst of this mess, but since the ordering issue is not obvious, I think it's worth calling out explicitly that the client may choose not to store the settings if things were received in the wrong order.

A server can remember the settings that it advertised, or store an
integrity-protected copy of the values in the ticket and recover the information
when accepting 0-RTT data. A server uses the HTTP/3 settings values in
determining whether to accept 0-RTT data.  If the server cannot determine that
the settings remembered by a client are compatible with its current settings, it
MUST NOT accept 0-RTT data.  (Remembered settings are compatible if a client

> complying with those settings would not violate the server's current settings.)

Reviewer 4: (I expect our implementation will just require an exact byte-for-byte match. It saves adding a funny callback and I wouldn't expect the values to actually change much. Same for the QUIC transport parameters.)

Reviewer 2: Storing a version number in place of storing settings is entirely appropriate.

Reviewer 4: Sure, that's a plausible optimization. Although that's orthogonal to whether you believe in equality or a weaker comparison.

(Whether the optimization is worth it, I'm not sure. The benefit is slightly smaller tickets. The cost is you need to make sure the meaning of your version number is consistent, which is a hidden sharp edge to deployment.)

A server MAY accept 0-RTT and subsequently provide different settings in its
SETTINGS frame. If 0-RTT data is accepted by the server, its SETTINGS frame MUST
NOT reduce any limits or alter any values that might be violated by the client
with its 0-RTT data.  The server MUST include all settings which differ from

> their default values.

Reviewer 4: What should the client do if it sees the server reduce limits or otherwise alter values?
What if it sees the server do so implicitly by omitting a non-default setting that was previously set in the remembered SETTINGS frame?

Author: Good point.  We currently don't require clients to police that, but it falls in the bucket of trivially-detectable protocol violations.

For clients using a 1-RTT QUIC connection, the initial value of each server
setting is the default value. When a 0-RTT QUIC connection is being used, the
initial value of each server setting is the value used in the previous session.
Clients SHOULD store the settings the server provided in the session being

> resumed and MUST comply with stored settings until the current server settings

Reviewer 5: I might be confused, but how can you have `MUST comply with stored` if storing is a `SHOULD`?

Reviewer 2: Yeah, this complicates the text somewhat.  If the intent is to have clients comply with either defaults or whatever they remember, then maybe:

Clients SHOULD store the settings the server provided in the connection where resumption information was provided.  A client MUST comply with either stored settings - or default values if no values are stored - when attempting 0-RTT.  Once a server has provided new settings, clients MUST comply with those values.

are received.  If the client did not store, or did not receive, server settings
from the previous session, the initial value of each setting is the default

> value.  A client can use these initial values to send requests before the

Reviewer 2: This business about defaults can be split into a new paragraph, ideally one preceding this more complicated part.

> Settings assume default values prior to receiving a SETTINGS frame, or if the SETTINGS frame does not include a value for the setting.  Endpoints can use these default values to send requests before receiving the a SETTINGS frame.

server's SETTINGS frame has arrived.  This removes the need for a client to wait
for the SETTINGS frame before sending requests.

A server can remember the settings that it advertised, or store an
integrity-protected copy of the values in the ticket and recover the information
when accepting 0-RTT data. A server uses the HTTP/3 settings values in

> determining whether to accept 0-RTT data.  If the server cannot recover the

Reviewer 5: Can we elaborate on what `recover` means? Does assuming the defaults count as `recover`? Also, how can the server recover them if it doesn't know if the client stored them or not?

> settings from a session being resumed, it MUST NOT accept 0-RTT data.

Reviewer 3: by my reading, the paragraph describes to methods: remember or recover. This new clause seems to imply if you can't recover, then remember is not allowed.

Is it more aligned with the intent to say that when a server chooses the recover method, and encodes the data in a ticket, and a client presents a ticket, then the server MUST recover the settings or else reject 0-RTT?

Reviewer 2: I don't think that we're mandating the use of session tickets, even if that is the most common method of implementing these.

I think that "recover" is probably still right, though it might be clearer here if we say "If a server cannot determine that the settings remembered by a client are compatible with its current settings, it MUST reject 0-RTT."

Then we need to define what "compatible" means.

A server MAY accept 0-RTT and subsequently provide different settings in its
SETTINGS frame. If 0-RTT data is accepted by the server, its SETTINGS frame MUST
NOT reduce any limits or alter any values that might be violated by the client
with its 0-RTT data.  The server MUST include all settings which differ from

> their default values, even if the value is unchanged from the previous session.

Reviewer 2: I might drop the bit from the comma onwards.  The only case we have two values is in 0-RTT (not resumption, as this implies) and that text is now very clear that the remembered values only apply until SETTINGS arrives.
